pkg/rancher-desktop/main/diagnostics/wslFromStore.ts

```diff
@@ -0,0 +1,221 @@
+        const stdout: string = ex.stdout;
+
+        if (stdout.includes(failureMatch)) {
+          return stdout;
+        }
+      }
+      throw ex;
+    }
+  }
+
+  _cachedWSLHelpText: string | undefined;
+  async getWSLHelpText(): Promise<string> {
+    if (this._cachedWSLHelpText) {
+      return this._cachedWSLHelpText;
+    }
+    // `wsl.exe --help` always exits with -1; that's okay as long as the output
+    // contains `wsl.exe` somewhere.
+    this._cachedWSLHelpText = await this.runWSL(['--help'], 'wsl.exe');
+
+    return this._cachedWSLHelpText;
+  }
+
+  async isKernelInstalled(): Promise<boolean> {
+    if ((await this.getWSLHelpText()).includes('--version')) {
+      const versions = await this.runWSL(['--version']);
+      const kernel = versions.split(/\r?\n/).map(s => s.trim()).find(v => /kernel/i.test(v));
+
+      console.debug(`Running WSL with ${ kernel || '(no kernel found)' }`);
+
+      return !!kernel;
+    } else {
+      try {
+        const status = await this.runWSL(['--status']);
+        const match = /:\s*((?:\d+\.){2,}\d+)/.exec(status);
+
+        if (match) {
+          console.debug(`Running WSL with kernel ${ match.groups?.[1] }`);
+
+          return true;
+        }
+        console.debug('`wsl --status` does not contain version string.');
+      } catch (ex) {
+        console.debug(ex);
+      }
+
+      return false;
+    }
+  }
+
+  async check() {
+    const storeURL = 'ms-windows-store://pdp/?ProductId=9P9TQF7MRM4R&mode=mini';
+
+    if (!(await this.getWSLHelpText()).includes('--exec')) {
+      // WSL is not installed
+      const description = 'Windows Subsystem for Linux is not installed.';
+      const fixes: { description: string, url?: string }[] = [];
+
+      if (await this.canUseStoreWSL()) {
+        fixes.push({
+          description: `Install Windows Subsystem for Linux from the [Microsoft Store](${ storeURL }).`,
+          url:         storeURL,
+        });
+      } else {
+        fixes.push({
+          description: `Install Windows Subsystem for Linux manually.`,
+          url:         'https://aka.ms/wslinstall',
+        });
+      }
+
+      return {
+        passed: false, description, fixes,
+      };
+    }
+
+    if (!await this.isKernelInstalled()) {
+      // The kernel is not installed; this covers virtualization not available.
+      return {
+        passed:      false,
+        description: `The WSL kernel does not appear to be installed.`,
+        fixes:       [{ description: 'Install the WSL kernel with `wsl.exe --update`' }],
+      };
+    }
+
+    if (await this.canUseStoreWSL()) {
+      const description = await this.checkWSLFromStore();
+
+      if (description) {
+        // WSL is installed, but we're using the in-box version.
+        return {
+          passed: false,
+          description,
+          fixes:  [{
+            description: `Install Windows Subsystem for Linux from the [Microsoft Store](${ storeURL }).`,
```

Comment:
Is this really a failure though? Doesn't Rancher Desktop work fine with the installed version; we are just letting them know that a newer version is available from the store?

I guess the problem is that if this is a passing diagnostic, then user will simply not see it. So this would be another example of wanting to differentiate between "info" and "warning" diagnostics?

Reply:
Yep, currently there's only pass or fail, so this would have to be fail (and the user can mute it if they don't care).